### ISSUE TITLE
[CODE IMPROVEMENT] Avoid gradient sync during gradient accumulation

### ISSUE THREAD

**Issue by rohitgr7** [h2oai/h2o-llmstudio#664]
### 🔧 Proposed code refactoring

Gradient sync happens during `loss.backward` which is not required if currently it's just accumulating on GPUs and we don't have to sync them on each iteration. This is required only before `optimizer.step`. So this can be blocked using:
https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync

### Motivation

Avoid unnecessary communication. Might get speed-ups.
